spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/DatastoreEntityConverterImpl.java

```diff
@@ -0,0 +1,140 @@
+ *
+ * @author Dmitry Solomakha
+ *
+ * @since 1.1
+ */
+public class DatastoreEntityConverterImpl implements DatastoreEntityConverter {
+	private DatastoreMappingContext mappingContext;
+
+	private EntityInstantiators instantiators = new EntityInstantiators();
+
+	private static Map<Class<?>, Function<?, Value<?>>> valFactories;
+
+	static {
+		ImmutableMap.Builder<Class<?>, Function<?, Value<?>>> builder = new ImmutableMap.Builder<>();
+
+		builder.put(Blob.class, (Function<Blob, Value<?>>) BlobValue::of);
+		builder.put(Boolean.class, (Function<Boolean, Value<?>>) BooleanValue::of);
+		builder.put(Double.class, (Function<Double, Value<?>>) DoubleValue::of);
+		builder.put(Entity.class, (Function<Entity, Value<?>>) EntityValue::of);
+		builder.put(Key.class, (Function<Key, Value<?>>) KeyValue::of);
+		builder.put(LatLng.class, (Function<LatLng, Value<?>>) LatLngValue::of);
+		builder.put(Long.class, (Function<Long, Value<?>>) LongValue::of);
+		builder.put(String.class, (Function<String, Value<?>>) StringValue::of);
+		builder.put(Timestamp.class, (Function<Timestamp, Value<?>>) TimestampValue::of);
+
+		valFactories = builder.build();
+	}
+
+	DatastoreEntityConverterImpl(DatastoreMappingContext mappingContext) {
+		this.mappingContext = mappingContext;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public <R> R read(Class<R> aClass, Entity entity) {
+		DatastorePersistentEntity<R> persistentEntity = (DatastorePersistentEntity<R>) this.mappingContext
+				.getPersistentEntity(aClass);
+
+		EntityPropertyValueProvider propertyValueProvider = new EntityPropertyValueProvider(entity);
+
+		ParameterValueProvider<DatastorePersistentProperty> parameterValueProvider =
+				new PersistentEntityParameterValueProvider<>(persistentEntity, propertyValueProvider, null);
+
+		EntityInstantiator instantiator = this.instantiators.getInstantiatorFor(persistentEntity);
+		R instance = instantiator.createInstance(persistentEntity, parameterValueProvider);
+		PersistentPropertyAccessor accessor = persistentEntity.getPropertyAccessor(instance);
+
+		persistentEntity.doWithProperties(
+				(PropertyHandler<DatastorePersistentProperty>) datastorePersistentProperty -> {
+					Object value = propertyValueProvider.getPropertyValue(datastorePersistentProperty);
+					accessor.setProperty(datastorePersistentProperty, value);
+				});
+
+		return instance;
+	}
+
+	@Override
+	public void write(Object source, Entity.Builder sink) {
+		DatastorePersistentEntity<?> persistentEntity = this.mappingContext
+				.getPersistentEntity(source.getClass());
+		PersistentPropertyAccessor accessor = persistentEntity
+				.getPropertyAccessor(source);
+		persistentEntity.doWithProperties(
+				(DatastorePersistentProperty datastorePersistentProperty) ->
+					writeProperty(sink, accessor, datastorePersistentProperty)
+				);
+	}
+
+	@SuppressWarnings("unchecked")
+	private <T> void writeProperty(Entity.Builder sink, PersistentPropertyAccessor accessor,
```

Comment:
Because this still required the unchecked suppress line, I think we should get rid of the type param T because it is just verbose now. 

You can change the following lines:

`Function valueFactory = valFactories.get(persistentProperty.getType());`

`Value val = (Value) valueFactory.apply(propertyVal);`

Getting rid of the `T` and the `?` is fine I think. We were just casting with those anyway.